File: apps/web/src/engine/executors/client-safe/ConditionExecutor.ts
/**
 * Condition node executor - handles boolean logic and branching
 */

import { 
  Node, 
  ExecutorResult, 
  ExecutorValidation,
  ExecutionContext as TypedExecutionContext
} from '@/shared/types/core';

import { ClientSafeExecutor } from '../base-executor';

export class ConditionExecutor extends ClientSafeExecutor {
  async validateInputs(node: Node, context: TypedExecutionContext): Promise<ExecutorValidation> {
    const commonValidation = this.validateCommonInputs(node);
    if (!commonValidation.isValid) {
      return commonValidation;
    }

    const errors: string[] = [];

    // Check condition type
    const conditionType = this.getNodeProperty(node, 'conditionType', 'expression');
    
    if (conditionType === 'expression') {
      // Validate condition expression
      const condition = this.getNodeProperty(node, 'expression', '');
      if (!condition) {
        errors.push('Condition expression is required');
      }
    } else if (conditionType === 'max_iterations') {
      // No specific validation needed for max_iterations type
    } else {
      errors.push(`Invalid condition type: ${conditionType}`);
    }

    return {
      isValid: errors.length === 0,
      errors
    };
  }

  async execute(node: Node, context: TypedExecutionContext, options?: any): Promise<ExecutorResult> {
    const conditionType = this.getNodeProperty(node, 'conditionType', 'expression') as string;
    const inputs = this.getInputValues(node, context);
    
    try {
      let result: boolean;
      
      if (conditionType === 'max_iterations') {
        // Check if any preceding nodes have reached max iterations
        result = this.checkPrecedingNodesMaxIterations(node, context);
      } else {
        // Default to expression evaluation
        const expression = this.getNodeProperty(node, 'expression', '');
        result = this.evaluateCondition(expression, inputs, context);
      }
      
      return this.createSuccessResult(result, 0, {
        conditionType,
        inputs,
        evaluatedAt: new Date().toISOString()
      });
    } catch (error) {
      throw this.createExecutionError(
        `Failed to evaluate condition: ${error instanceof Error ? error.message : String(error)}`,
        node,
        { conditionType, inputs, error: error instanceof Error ? error.message : String(error) }
      );
    }
  }

  /**
   * Evaluate a condition expression
   */
  private evaluateCondition(
    condition: string, 
    inputs: Record<string, any>, 
    context: TypedExecutionContext
  ): boolean {
    // Create evaluation context with inputs and flattened node outputs
    const evaluationContext: Record<string, any> = {
      ...inputs,
      executionCount: context.nodeExecutionCounts
    };

    // Flatten node outputs to make their properties directly accessible
    for (const [nodeId, nodeOutput] of Object.entries(context.nodeOutputs)) {
      if (typeof nodeOutput === 'object' && nodeOutput !== null) {
        // Spread object properties to make them accessible (e.g., {counter: 1} -> counter: 1)
        Object.assign(evaluationContext, nodeOutput);
      } else {
        // For primitive values, use the node ID as key
        evaluationContext[nodeId] = nodeOutput;
      }
    }

    console.log(`[ConditionExecutor] Evaluating "${condition}" with context:`, evaluationContext);

    // Simple expression evaluation
    // In production, you'd want to use a more robust expression evaluator
    return this.evaluateExpression(condition, evaluationContext);
  }

  /**
   * Simple expression evaluator for conditions
   */
  private evaluateExpression(expression: string, context: Record<string, any>): boolean {
    // Replace variables in the expression
    let evaluatedExpression = expression;
    
    for (const [key, value] of Object.entries(context)) {
      const regex = new RegExp(`\\b${key}\\b`, 'g');
      const stringValue = typeof value === 'string' ? `"${value}"` : String(value);
      evaluatedExpression = evaluatedExpression.replace(regex, stringValue);
    }

    // Handle basic comparisons and boolean logic
    try {
      // This is a simplified evaluator - in production use a proper expression parser
      return this.evaluateSimpleBoolean(evaluatedExpression);
    } catch (error) {
      console.warn(`Failed to evaluate expression "${expression}":`, error);
      return false;
    }
  }

  private evaluateSimpleBoolean(expression: string): boolean {
    const cleaned = expression.trim();
    
    // Handle boolean literals
    if (cleaned === 'true') return true;
    if (cleaned === 'false') return false;
    
    // Handle simple comparisons
    const comparisonOps = ['===', '!==', '==', '!=', '<=', '>=', '<', '>'];
    for (const op of comparisonOps) {
      if (cleaned.includes(op)) {
        const parts = cleaned.split(op).map(s => s.trim());
        if (parts.length === 2) {
          return this.compareValues(parts[0]!, parts[1]!, op);
        }
      }
    }

    // Handle logical operators
    if (cleaned.includes('&&')) {
      const parts = cleaned.split('&&').map(s => s.trim());
      return parts.every(part => this.evaluateSimpleBoolean(part));
    }

    if (cleaned.includes('||')) {
      const parts = cleaned.split('||').map(s => s.trim());
      return parts.some(part => this.evaluateSimpleBoolean(part));
    }

    return false;
  }

  private compareValues(left: string, right: string, operator: string): boolean {
    const leftVal = this.parseValue(left || '');
    const rightVal = this.parseValue(right || '');

    switch (operator) {
      case '===':
      case '==':
        return leftVal === rightVal;
      case '!==':
      case '!=':
        return leftVal !== rightVal;
      case '<':
        return leftVal < rightVal;
      case '<=':
        return leftVal <= rightVal;
      case '>':
        return leftVal > rightVal;
      case '>=':
        return leftVal >= rightVal;
      default:
        return false;
    }
  }

  private parseValue(value: string): any {
    const trimmed = value.trim();
    
    // Remove quotes for strings
    if ((trimmed.startsWith('"') && trimmed.endsWith('"')) ||
        (trimmed.startsWith("'") && trimmed.endsWith("'"))) {
      return trimmed.slice(1, -1);
    }

    // Parse numbers
    const num = Number(trimmed);
    if (!isNaN(num)) {
      return num;
    }

    // Parse booleans
    if (trimmed === 'true') return true;
    if (trimmed === 'false') return false;

    return trimmed;
  }

  /**
   * Check if any preceding nodes have reached their max iterations
   */
  private checkPrecedingNodesMaxIterations(node: Node, context: TypedExecutionContext): boolean {
    // Get all incoming nodes
    const incomingArrows = context.incomingArrows[node.id] || [];
    
    // Check all directly connected source nodes
    for (const arrow of incomingArrows) {
      const sourceNodeId = arrow.source;
      if (!sourceNodeId) continue;
      
      const sourceNode = context.nodesById[sourceNodeId];
      if (!sourceNode) continue;
      
      // Check if this node has max iterations defined and has reached them
      const executionCount = context.nodeExecutionCounts[sourceNodeId] || 0;
      const maxIterations = sourceNode.data?.iterationCount || sourceNode.data?.maxIterations;
      
      if (maxIterations && executionCount >= maxIterations) {
        console.log(`[ConditionExecutor] Node ${sourceNodeId} has reached max iterations (${executionCount}/${maxIterations})`);
        return true;
      }
    }
    
    // Also check for any nodes that might be in a cycle with this condition node
    // Look for nodes that feed back to the same nodes this condition feeds to
    const outgoingArrows = context.outgoingArrows[node.id] || [];
    for (const outArrow of outgoingArrows) {
      const targetNodeId = outArrow.target;
      if (!targetNodeId) continue;
      
      // Find nodes that connect to this same target (potential loop participants)
      for (const [nodeId, nodeArrows] of Object.entries(context.outgoingArrows)) {
        if (nodeId === node.id) continue; // Skip self
        
        const nodeConnectsToSameTarget = nodeArrows.some(arrow => arrow.target === targetNodeId);
        if (nodeConnectsToSameTarget) {
          const loopNode = context.nodesById[nodeId];
          if (!loopNode) continue;
          
          const executionCount = context.nodeExecutionCounts[nodeId] || 0;
          const maxIterations = loopNode.data?.iterationCount || loopNode.data?.maxIterations;
          
          if (maxIterations && executionCount >= maxIterations) {
            console.log(`[ConditionExecutor] Loop participant ${nodeId} has reached max iterations (${executionCount}/${maxIterations})`);
            return true;
          }
        }
      }
    }
    
    // No nodes have reached max iterations
    return false;
  }
}

File: apps/server/src/engine/executors/condition_executor.py
"""
Condition node executor - handles boolean logic and branching
"""

from typing import Dict, Any, TYPE_CHECKING
import time
import re
import logging

from .base_executor import BaseExecutor, ExecutorResult
from .utils import get_input_values
from .validator import ValidationResult, validate_required_fields, validate_enum_field

if TYPE_CHECKING:
    from ..types import Ctx

logger = logging.getLogger(__name__)


class ConditionExecutor(BaseExecutor):
    """
    Condition node executor that evaluates boolean expressions or checks max iterations.
    Supports two modes:
    - expression: Evaluates a boolean expression with variables
    - max_iterations: Checks if all nodes with max iterations have reached their limit
    """
    
    async def validate(self, node: Dict[str, Any], context: 'Ctx') -> ValidationResult:
        """Validate condition node configuration"""
        errors = []
        warnings = []
        
        properties = node.get("properties", {})
        condition_type = properties.get("conditionType", "expression")
        
        # Validate condition type enum
        enum_error = validate_enum_field(
            properties, 
            "conditionType", 
            ["expression", "detect_max_iterations"],
            case_sensitive=True
        )
        if enum_error:
            errors.append(enum_error)
        
        if condition_type == "expression":
            # Validate required expression field
            field_errors = validate_required_fields(
                properties,
                ["expression"],
                {"expression": "Condition expression"}
            )
            errors.extend(field_errors)
            
            # Basic expression validation
            expression = properties.get("expression", "")
            if expression and not any(op in expression for op in ["==", "!=", "<", ">", "<=", ">=", "and", "or", "true", "false"]):
                warnings.append("Expression may not contain valid comparison operators")
        
        elif condition_type == "detect_max_iterations":
            # No specific validation needed for max_iterations type
            pass
        
        return ValidationResult(
            is_valid=len(errors) == 0,
            errors=errors,
            warnings=warnings
        )
    
    async def execute(self, node: Dict[str, Any], context: 'Ctx') -> ExecutorResult:
        """Execute condition node and return boolean result"""
        start_time = time.time()
        
        properties = node.get("properties", {})
        condition_type = properties.get("conditionType", "expression")
        inputs = get_input_values(node, context)
        
        try:
            if condition_type in ["detect_max_iterations"]:
                result = self._check_preceding_nodes_max_iterations(node, context)
            else:
                # Default to expression evaluation
                expression = properties.get("expression", "")
                result = self._evaluate_condition(expression, inputs, context)
            
            execution_time = time.time() - start_time
            
            # Pass through input data while storing condition result in metadata
            # If there's only one input, pass it directly; otherwise pass the full inputs dict
            if len(inputs) == 1:
                output_data = next(iter(inputs.values()))
            elif len(inputs) > 1:
                output_data = inputs
            else:
                # No inputs, just pass through empty dict
                output_data = {}
            
            return ExecutorResult(
                output=output_data,
                metadata={
                    "conditionType": condition_type,
                    "conditionResult": result,  # Store the boolean result for flow control
                    "inputs": inputs,
                    "evaluatedAt": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
                    "executionTime": execution_time
                },
                execution_time=execution_time
            )
        
        except Exception as e:
            # Pass through input data even on error, but set condition result to False
            if len(inputs) == 1:
                output_data = next(iter(inputs.values()))
            elif len(inputs) > 1:
                output_data = inputs
            else:
                output_data = {}
            
            return ExecutorResult(
                output=output_data,
                error=f"Failed to evaluate condition: {str(e)}",
                metadata={
                    "conditionType": condition_type,
                    "conditionResult": False,  # Default to False on error
                    "inputs": inputs,
                    "error": str(e)
                },
                execution_time=time.time() - start_time
            )
    
    def _evaluate_condition(self, expression: str, inputs: Dict[str, Any], context: 'Ctx') -> bool:
        """Evaluate a condition expression"""
        # Create evaluation context with inputs and flattened node outputs
        evaluation_context = {**inputs}
        
        # Add execution counts
        evaluation_context["executionCount"] = context.exec_cnt
        
        # Flatten node outputs to make their properties directly accessible
        for node_id, node_output in context.outputs.items():
            if isinstance(node_output, dict):
                # Spread object properties to make them accessible
                evaluation_context.update(node_output)
            else:
                # For primitive values, use the node ID as key
                evaluation_context[node_id] = node_output
        
        logger.debug(f"[ConditionExecutor] Evaluating '{expression}' with context: {evaluation_context}")
        
        # Simple expression evaluation
        return self._evaluate_expression(expression, evaluation_context)
    
    def _evaluate_expression(self, expression: str, context: Dict[str, Any]) -> bool:
        """Simple expression evaluator for conditions"""
        # Replace variables in the expression
        evaluated_expression = expression
        
        # Replace variable references with their values
        for key, value in context.items():
            # Use word boundaries to avoid partial matches
            pattern = rf'\b{re.escape(key)}\b'
            if isinstance(value, str):
                replacement = f'"{value}"'
            else:
                replacement = str(value)
            evaluated_expression = re.sub(pattern, replacement, evaluated_expression)
        
        # Also handle {{variable}} syntax
        def replace_template(match):
            var_name = match.group(1)
            value = context.get(var_name)
            if value is None:
                return "None"
            elif isinstance(value, str):
                return f'"{value}"'
            else:
                return str(value)
        
        evaluated_expression = re.sub(r'\{\{(\w+)\}\}', replace_template, evaluated_expression)
        
        logger.debug(f"[ConditionExecutor] Evaluated expression: {evaluated_expression}")
        
        try:
            # Convert to Python boolean operators
            evaluated_expression = evaluated_expression.replace(" and ", " and ")
            evaluated_expression = evaluated_expression.replace(" or ", " or ")
            evaluated_expression = evaluated_expression.replace("&&", " and ")
            evaluated_expression = evaluated_expression.replace("||", " or ")
            evaluated_expression = evaluated_expression.replace("===", "==")
            evaluated_expression = evaluated_expression.replace("!==", "!=")
            
            # Use eval with restricted context for safety
            # In production, use a proper expression parser like simpleeval
            result = eval(evaluated_expression, {"__builtins__": {}}, {})
            return bool(result)
        except Exception as e:
            logger.warning(f"Failed to evaluate expression '{expression}': {e}")
            return False
    
    def _check_preceding_nodes_max_iterations(self, node: Dict[str, Any], context: 'Ctx') -> bool:
        """Check if all nodes with max iterations have reached their limit"""
        # Get all incoming nodes
        node_id = node["id"]
        incoming_arrows = context.graph.incoming.get(node_id, [])
        
        # Track if all nodes that have max iterations defined have reached them
        has_nodes_with_max_iterations = False
        all_max_iterations_reached = True
        
        # Check all directly connected source nodes
        for arrow in incoming_arrows:
            source_node_id = arrow.source
            if not source_node_id:
                continue
            
            source_node = context.graph.nodes.get(source_node_id)
            if not source_node:
                continue
            
            # Check if this node has max iterations defined
            execution_count = context.exec_cnt.get(source_node_id, 0)
            max_iterations = source_node.max_iter
            
            if max_iterations:
                has_nodes_with_max_iterations = True
                # Check if node was skipped due to max iterations
                if source_node_id in context.skipped and context.skipped.get(source_node_id) == "max_iterations":
                    logger.debug(f"[ConditionExecutor] Node {source_node_id} was skipped due to max iterations ({execution_count}/{max_iterations})")
                elif execution_count < max_iterations:
                    all_max_iterations_reached = False
                    logger.debug(f"[ConditionExecutor] Node {source_node_id} has NOT reached max iterations yet ({execution_count}/{max_iterations})")
                else:
                    logger.debug(f"[ConditionExecutor] Node {source_node_id} has reached max iterations ({execution_count}/{max_iterations})")
        
        # Also check for any nodes that might be in a cycle with this condition node
        outgoing_arrows = context.graph.outgoing.get(node_id, [])
        for out_arrow in outgoing_arrows:
            target_node_id = out_arrow.target
            if not target_node_id:
                continue
            
            # Find nodes that connect to this same target (potential loop participants)
            for check_node_id, node_arrows in context.graph.outgoing.items():
                if check_node_id == node_id:
                    continue  # Skip self
                
                node_connects_to_same_target = any(
                    arrow.target == target_node_id for arrow in node_arrows
                )
                
                if node_connects_to_same_target:
                    loop_node = context.graph.nodes.get(check_node_id)
                    if not loop_node:
                        continue
                    
                    execution_count = context.exec_cnt.get(check_node_id, 0)
                    max_iterations = loop_node.max_iter
                    
                    if max_iterations:
                        has_nodes_with_max_iterations = True
                        # Check if node was skipped due to max iterations
                        if check_node_id in context.skipped and context.skipped.get(check_node_id) == "max_iterations":
                            logger.debug(f"[ConditionExecutor] Loop participant {check_node_id} was skipped due to max iterations ({execution_count}/{max_iterations})")
                        elif execution_count < max_iterations:
                            all_max_iterations_reached = False
                            logger.debug(f"[ConditionExecutor] Loop participant {check_node_id} has NOT reached max iterations yet ({execution_count}/{max_iterations})")
                        else:
                            logger.debug(f"[ConditionExecutor] Loop participant {check_node_id} has reached max iterations ({execution_count}/{max_iterations})")
        
        # Return true only if we found nodes with max iterations AND all of them have reached their limit
        return has_nodes_with_max_iterations and all_max_iterations_reached